functions.cmake

```diff
@@ -228,13 +233,31 @@ function(_add_library)
       add_library(${ARGV} ${VERSION_FILE_PATH})
     else()
       add_library(${ARGV})
+      set_target_properties(
+        ${ARGV0} 
+        PROPERTIES
+        VS_GLOBAL_CL_MPCount ${CLR_CMAKE_CPU_CORES}
+      )
     endif(NOT WIN32)
     list(FIND CLR_CROSS_COMPONENTS_LIST ${ARGV0} INDEX)  
     if (DEFINED CLR_CROSS_COMPONENTS_LIST AND ${INDEX} EQUAL -1)  
      set_target_properties(${ARGV0} PROPERTIES EXCLUDE_FROM_ALL 1)  
     endif()  
 endfunction()
 
+function(_add_custom_target)
+    if(NOT WIN32)
+      add_custom_target(${ARGV})
+    else()
+      add_custom_target(${ARGV})
+      set_target_properties(
+        ${ARGV0} 
+        PROPERTIES
```

Comment:
My Cmake skills are some what limited but it would seem like there should be a way to add this property globally instead of requiring this wrapper. Is that not true?

Reply:
Ideally something like if WIN32 set this global property.

Reply:
The problem is that it is so called target property so it can only be set for a given target. I have tried to set it as global property but it had no effect on generated projects. It specifically defined as a property which can be set on targets in CMake documentation:

https://cmake.org/cmake/help/v3.9/manual/cmake-properties.7.html

Reply:
There is nothing like that for target properties like this one. However, googling a bit out of interest has revealed there was something that can help added to CMake in version 3.7 -  BUILDSYSTEM_TARGETS property that contains a list of all targets. So if we moved to this version of cmake as the minimum for Windows, we could use that at the end of the main CMakeFile.txt to set the property for all targets.

Reply:
In general CMake is not very expressive in Visual Studio generators and perhaps it would necessary to submit to their repo a PR with general feature enabling custom modifications of vcxprojects like:
- VS_GLOBAL_NODE_<variable> -> accepts complex values
- with value set to arguments of EmitNode generator function arguments
- EmitNode(string & nodeName, string* attributeNames, string* attributeValues, string & nodeValue, string & parentNodeXPath);

Reply:
I have made some preliminary tests on CMake 3.9 with BUILDSYSTEM_TARGETS before going with this solution and it's value was always empty at the top level scope. Additional test inside _add_library call which added target were negative as well - BUILDSYSTEM_TARGETS was empty despite that target existed since it was added one code line earlier. I have not investigated it any further.

Reply:
Please note it is not a variable, but a directory property. Maybe you know and read it that way, but just in case you didn't ...

Reply:
Oh my, I used it as a variable. Reading documentation again and checking it's usage - at first glance it seems the problem could be with a need to efficiently retrieve all subdirectories before enumerating and reading BUILDSYSTEM_TARGETS value from each of them and than setting targets properties - researching it now.

Reply:
That's why I've said in one of my previous responses that we should use it at the end of the root CMakeLists.txt. At that point, all subdirectories were already processed.

Reply:
Yeah, I have noticed that in your comment and even did that when tested usage of this property -
 nevertheless I have used it as a variable .... not directory property.

Reply:
@4creators as an alternative were you able to find any way to have CMake import a custom msbuild targets/props file into all projects? If so then we could put whatever global properties we want in there and just always have it imported.

Reply:
@weshaggard 
It is my favorite solution to the problem but unfortunately it is impossible according to my knowledge - I have described this in PR description:

> Researching CMake support for generation of custom properties/nodes in MSBuild
project files I have found to viable implementation options:

> use VS_GLOBAL_

> use add_compile_options(/MPx) where x is expected number of CL processess

This why I have suggested to create such feature in CMake and commit it to upstream product see https://github.com/dotnet/coreclr/pull/14578#discussion_r145780476

Reply:
Given we plan to move to MSBuild15 only soon another option is to bypass CMake for this matter completely and just use an msbuild extension point. See https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.Common.props#L36 which will import the closest Directory.Build.props file that if finds up the directory tree. Given these projects are generated under the bin\ directory you should be able to put one in the root of the coreclr repo and it will automatically get imported. This should give you the flexibility to set the properties you need without making any CMake changes.